language/rust/language.go

```diff
@@ -0,0 +1,272 @@
+		if err != nil {
+			return nil, pkgerrors.WithStack(err)
+		}
+		defer func() {
+			err = errors.Join(file.Close(), err)
+		}()
+
+		scanner := bufio.NewScanner(file)
+		for line := 0; scanner.Scan(); line++ {
+			if err := scanner.Err(); err != nil {
+				return nil, pkgerrors.WithStack(err)
+			}
+
+			if strings.Contains(scanner.Text(), "#[cfg(test)]") {
+				linePerFile[filePath] = line
+
+				break
+			}
+		}
+	}
+
+	return linePerFile, nil
+}
+
+// ExecuteTests invokes the language specific testing on the given repository.
+func (l *Language) ExecuteTests(logger *log.Logger, repositoryPath string) (testResult *language.TestResult, problems []error, err error) {
+	// HACK Tests in Rust are within the implementation file, but excluding coverage more granular than file-level is currently unstable (https://github.com/rust-lang/rust/issues/84605). Therefore, we assume that test are always at the end of the file and ignore reported coverage after the "#[cfg(test)]" compiler directive.
+	testStartLinePerFile, err := l.testDirectiveLinePerFile(logger, repositoryPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	logger.Info("rust test locations", "locations", testStartLinePerFile)
+
+	ctx, cancel := context.WithTimeout(context.Background(), language.DefaultExecutionTimeout)
+	defer cancel()
+	coverageFilePath := filepath.Join(repositoryPath, "coverage.json")
+	commandOutput, err := util.CommandWithResult(ctx, logger, &util.Command{
+		Command: []string{
+			tools.SymflowerPath, "test",
+			"--language", "rust",
+			"--workspace", repositoryPath,
+			"--coverage-file", coverageFilePath,
+		},
+
+		Directory: repositoryPath,
+	})
+	if err != nil {
+		return nil, nil, pkgerrors.WithMessage(pkgerrors.WithStack(err), commandOutput)
+	}
+
+	testsTotal, testsPass, e := parseSymflowerTestOutput(commandOutput)
+	if e != nil {
+		problems = append(problems, pkgerrors.WithMessage(pkgerrors.WithStack(e), commandOutput))
+	}
+	// If there are test failures, then this is just a soft error since we still are able to receive coverage data.
+	if err != nil {
+		if testsTotal-testsPass <= 0 {
+			return nil, nil, pkgerrors.WithMessage(pkgerrors.WithStack(err), commandOutput)
+		}
+
+		problems = append(problems, pkgerrors.WithMessage(pkgerrors.WithStack(err), commandOutput))
+	}
+
+	testResult = &language.TestResult{
+		TestsTotal: uint(testsTotal),
+		TestsPass:  uint(testsPass),
+
+		StdOut: commandOutput,
+	}
+
+	coverageData, err := language.ParseCoverage(logger, coverageFilePath)
+	if err != nil {
+		return testResult, problems, pkgerrors.WithMessage(pkgerrors.WithStack(err), commandOutput)
+	}
+	for _, block := range coverageData {
+		// HACK The coverage should only contain relative paths but contains a weird build folder on macOS.
+		if index := strings.Index(block.FilePath, "src"); index != -1 && osutil.IsDarwin() && filepath.IsAbs(block.FilePath) {
+			block.FilePath = block.FilePath[index:]
+		}
```

Comment:
TODO Get rid of this hack.

Reply:
Fixed. https://github.com/symflower/eval-dev-quality/pull/452/commits/1c446da29e320446b54bbfb0ce633df1d7cd1e5c